---
title: "GEO-submitter"
runtime: shiny
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
    source_code: embed
    theme: bootstrap
---

```{r setup, include=FALSE}
library(flexdashboard)
library(highcharter)
library(NanoStringNorm)
library(ggplot2)
library(shiny)
library(shinyjs)
source("util/helpers.R")
source("util/import_rcc.R")
options(shiny.maxRequestSize = 9*1024^2) # 9mb limit
```

Upload {data-orientation=rows}
=====================================  

Inputs {.sidebar data-width=237}
-----------------------------------------------------------------------
```{r introduction}
h4('Introduction')
```
GEO-submitter is an easy to use shiny tool, developed to make submitting gene expression study data easier.

```{r upload bar}
# Data Type
tags$hr()
selectInput(inputId = 'data_type', 
            label = h4('Select Data Type'), 
            choices = c('Nanostring', "Microarray"), 
            selected = "Nanostring", 
            selectize = T)
tags$hr()

# Input Files
fileInput(inputId = 'upload', 
          label = h4('Choose Files to Upload'),
          multiple = T,
          accept = c(
            '.RCC',
            '.rcc',
            '.rcc.gz'
            )
      )
tags$hr()

# Normalization method
selectInput(inputId = "normalization", 
            label = h4('Normalization Method'), 
            choices = c("Housekeeping-Sum" = "housekeeping.sum",
                        "Housekeeping-Mean" = "housekeeping.geo.mean", 
                        "DESeq2" = "deseq2"), 
            selectize = T)
actionButton("norm", "More Info", width = "100px")
tags$hr()

# Norm modal for more info
observeEvent(input$norm, {
  showModal(
    modalDialog(
      title = "Normalization Methods",
      size = "m",
      "Below are a breakdown of the different methods used for normalizing Nanostring data.",
      easyClose = TRUE
    ))
})
```

```{r data reactive}
# Reactive values to store imported data
data <- reactive({
  
  # Err if NULL
  if(is.null(input$upload)){
    return(NULL)
  }
  data <- import_rcc2(input$upload$datapath)
  showNotification("Files processed...", type = "message")
  return(list(genes = data$x, metadata = data$header))
})
```

Application author: [Thomas W. Battaglia](https://github.com/twbattaglia)


Row
-----------------------------------------------------------------------

### Number of genes

```{r valuebox geneN}
output$geneNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data())){
    return(valueBox(0, caption = "Number of genes", icon = "fa-comments"))
  }
  geneN <- nrow(data()$genes)
  valueBox(geneN,caption = "Number of genes", icon = "fa-comments")
})
valueBoxOutput(outputId = "geneNumber")
```

### Number of samples

```{r valuebox sampleN}
output$sampleNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data())){
    return(valueBox(0, caption = "Number of samples", icon = "fa-comments"))
  }
  
  sampleN <- ncol(data()$metadata)
  valueBox(sampleN, caption = "Number of samples", icon = "fa-comments")
})
valueBoxOutput(outputId = "sampleNumber")
```


Row 
-------------------------------------
   
### Raw
  
```{r raw boxplot}
output$unnormalized <- renderHighchart({
  
  # Err if NULL
  if(is.null(data())){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Generating Plots...", type = "message")
  long_format <- subset(data()$genes, select = c(-1,-3)) %>%
    reshape2::melt(id.vars = c("Name")) 
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(long_format$value), 
                          by = long_format$variable,
                          color = "#2980b9")  %>%
    hc_title(text = "Raw Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    #hc_xAxis(categories = long_format$variable) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("unnormalized")
``` 
    
    
### Normalized

```{r norm boxplot}
output$normalized <- renderHighchart({
  # Err if NULL
  if(is.null(input$upload)){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Generating Plots...", type = "message")
  
  nsn <- NanoStringNorm::NanoStringNorm(
    x = data()$genes,
    anno = NA,
    CodeCount = 'sum',
    Background = 'mean',
    SampleContent = input$normalization,
    round.values = FALSE,
    take.log = FALSE,
    return.matrix.of.endogenous.probes = TRUE
  )
  
  # Convert to long format
  long_format <- reshape2::melt(nsn, id.vars = c("Name")) 
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(long_format$value), 
                          by = long_format$Var2,
                          color = "#2980b9")  %>%
    hc_title(text = "Normalized Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    #hc_xAxis(categories = long_format$Var2) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("normalized")
```



Experimental setup
===================================== 

Inputs {.sidebar data-width=237}
-----------------------------------------------------------------------
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed commodo bibendum imperdiet. Pellentesque pharetra hendrerit feugiat. Fusce ultrices malesuada purus, quis vehicula quam. Nulla facilisi. Phasellus pulvinar lacus id ante scelerisque eleifend. 

```{r save experiment}
actionButton(inputId = "saveExp", icon = icon("fa-list"), label = "Save Experiment", width = "215px")

# Store all reactive data assocaited with Series information
reactive <- reactiveValues()

observeEvent(input$saveExp, {
  showNotification("Saving data...", type = "message")
  
  # On save, add values to appropriate sections
  # TODO validations
  reactive$name <- isolate(input$name)
  reactive$title <- isolate(input$title)
  reactive$summary <- isolate(input$summary)
  reactive$design <- isolate(input$design)
  reactive$extraction <- isolate(input$extraction)
  reactive$organism <- isolate(input$organism)
  reactive$extraction <- isolate(input$extraction)
  reactive$molecule <- isolate(input$molecule)
  reactive$source_type <- isolate(input$source_type)
  reactive$platform <- isolate(input$platform)
  reactive$pubmed <- isolate(input$pubmed)
})

```


Column {.tabset .tabset-fade}
-------------------------------------

### Experimental 
    
```{r experimental info 1}
h4("Name")
textInput(inputId = "name", 
          label = "Provide an identifier for this entity. This identifier is used only as an internal reference within a given file. The identifier will not appear on final GEO records.", 
          width = "400px",
          value = "", 
          placeholder = "Enter text...")

tags$hr()

h4("Title")
textInput(inputId = "title", 
          label = "Provide a unique title that describes the overall study. Max 255 characters.", 
          value = "", 
          width = "400px",
          placeholder = "Enter text...")

tags$hr()

h4("Summary")
textareaInput(inputId = "summary", 
          label = "Summarize the goals and objectives of this study. 
          The abstract from the associated publication may be suitable.", 
          value = "", 
          placeholder = "Enter text...")

h4("Design")
textareaInput(inputId = "design", 
          label = "Provide a description of the experimental design. Indicate how many Samples are analyzed, if replicates are included, are there control and/or reference Samples", 
          value = "", 
          placeholder = "Enter text...")


h4("Extraction Protocol")
textareaInput(inputId = "extraction", 
          label = "How samples were extracted", 
          value = "", 
          placeholder = "Enter text...")
```
   
### Platform 

```{r experimental info 2}
h4("Organism")
textInput(inputId = "organism", 
          label = "Identify the organism(s) from which the biological material was derived. See http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/", 
          value = "",
          width = "400px",
          placeholder = "Mus musculus")

tags$hr()

h4("Molecule Type")
selectInput(inputId = 'molecule', 
            label = 'Specify the type of molecule that was extracted from the biological material.', 
            choices = c("total RNA", "polyA RNA", "cytoplasmic RNA", "nuclear RNA", "genomic DNA", "protein", "other"),
            width = "400px",
            selectize=TRUE)

tags$hr()

h4("Sample Source")
textInput(inputId = "source_type", 
          label = "Enter sample source type.", 
          value = "", 
          width = "400px",
          placeholder = "Terminal Ileum")

tags$hr()

h4("Platform")
textInput(inputId = "platform", 
          label = "Enter the NCBI platform identifier used. See http://www.ncbi.nlm.nih.gov/geo/browse/?view=platforms&tool=findplatform to find the corresponding identifier.", 
          value = "", 
          width = "400px",
          placeholder = "GPL19964")

tags$hr()

h4("Pubmed ID (Optional)")
textInput(inputId = "pubmed", 
          label = "Optionally, specify a valid PubMed identifier (PMID) that references a published article describing this study. Most commonly, this information is not available at the time of submission - it can be added later once the data are published.", 
          value = "", 
          width = "400px",
          placeholder = "12345678")
```     

### Authors
    
```{r author input}
fluidRow(
    column(width = 4, 
           textInput(inputId = "firstName", 
                     label = h5("First Name"), 
                     value = "", 
                     placeholder = "Watson")),
    column(width = 4, 
           textInput(inputId = "middleInitial", 
                     label = h5("Middle Initial"), 
                     value = "", 
                     placeholder = "N")),
    column(width = 4, 
           textInput(inputId = "lastName", 
                     label = h5("Last Name"), 
                     value = "", 
                     placeholder = "Crick"),
           actionButton(inputId = "addAuthor", label = "Add Author", icon = icon("fa-list")),
           actionButton(inputId = "removeAuthor", label = "Remove Author", icon = icon("fa-list")))
    )
tags$hr()
```   


```{r author table}
reactive$authorDF <- data.frame(First = "",
                                Middle = "",
                                Last = "")

# Observe if 'add author' button is pressed
observeEvent(input$addAuthor, {
  showNotification("Adding Author...", type = "message")
  
  # Require first and last names
  req(input$firstName)
  req(input$lastName)
  
  # Add values to data frame (for showing)
  author_newLine <- isolate(c(input$firstName, input$middleName, input$lastName))
  isolate(reactive$authorDF <- rbind(as.matrix(reactive$authorDF), unlist(author_newLine)))
  
  # Concatenate names together for internal use only
  new_author <- isolate(paste(input$firstName, input$middleName, input$lastName, sep = ","))
  reactive$authorList <- c(isolate(reactive$authorList), new_author)
})

# Render table of authors
output$author_table <- renderTable({reactive$authorDF}, include.rownames = F)
tableOutput("author_table")
```



Export
=====================================

Inputs {.sidebar data-width=237}
-----------------------------------------------------------------------
The tool will export a Simple Omnibus Format in Text (SOFT) file, which will include all relavent information for a Minimum Information About a Microarray Experiment (MIAME) compliant experiment.


Row
-------------------------------------
    
### Author
    
```{r}
# Dev
output$authorlist <- renderPrint({ reactive$authorList })
verbatimTextOutput('authorlist')
```
 
### Chart 2
    
```{r}
output$name <- renderPrint({ reactive$name })
verbatimTextOutput('name')
``` 

Row
-------------------------------------
    
### Chart 3
    
```{r}
```
    
### Export

```{r}
output$downloadData <- downloadHandler(
  filename = function() { 
    paste("soft", '.txt', sep='') 
  },
  
  content = function(file) {
   # write table function
  }
)
downloadButton('downloadData', 'Download')
```




<style>
#sidebar.section.sidebar {
  background-color: white; 
  font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif !important;
}

.js-irs-0 .irs-bar {
border-top-color: #d01010;
border-bottom-color: #d01010;
} 

.js-irs-0 .irs-bar-edge {
border-color: #d01010;
}

.js-irs-0 .irs-single, .js-irs-0 .irs-bar-edge, .js-irs-0 .irs-bar {
background: #a00;
}

</style>
