---
title: "GEO-submitter"
runtime: shiny
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
    source_code: embed
    theme: bootstrap
---

```{r setup, include=FALSE}
library(flexdashboard)
library(shiny)
source("util/helpers.R")
source("util/import_rcc.R")
options(shiny.maxRequestSize = 9*1024^2) # 9mb limit
```


Introduction {data-orientation=rows}
=====================================  

Column
-------------------------------------
    
### Introduction
    
```{r introduction}
h4('Introduction')
```
GEO-submitter is an easy to use shiny tool, developed to make submitting gene expression study data easier.
Application author: [Thomas W. Battaglia](https://github.com/twbattaglia)

    
### Normalization Methods

```{r}
```



Upload {data-orientation=rows}
=====================================  

Inputs {.sidebar data-width=267}
-----------------------------------------------------------------------

```{r upload bar}
# Data Type
h4('Data Type')
selectizeInput(inputId = 'dataType', width = NULL,
               label = helpText("Select an array platform type."), 
               choices = c('Nanostring', "Microarray"), 
               options = list(placeholder = 'Select array type...     ', 
                              onInitialize = I('function() { this.setValue(""); }'))
               )
tags$hr()

# Input Files
h4('Upload Files')
fileInput(inputId = 'upload', 
          label = helpText("Use 'Shift' to select multiple files."), 
          multiple = T,
          accept = c(
            '.RCC',
            '.RCC.gz', 
            '.CEL',
            '.CEL.gz')
      )
tags$hr()

# Normalization method
conditionalPanel(condition = "input.dataType == 'Nanostring'",
                 h4('Normalization Method'),
                 selectInput(inputId = "normalizationNS", 
                             label = helpText("Select a normalization method"), 
                             choices = c("Housekeeping-Sum" = "housekeeping.sum",
                                         "Housekeeping-Mean" = "housekeeping.geo.mean", 
                                         "DESeq2" = "deseq2"), 
                             selectize = T),
                 tags$hr()
)

conditionalPanel(condition = "input.dataType == 'Microarray'",
                 h4('Normalization Method'),
                 selectInput(inputId = "normalizationMA", 
                             label = helpText("Select normalization method."), 
                             choices = c("RMA" = "RMA",
                                         "MAS5" = "MAS5", 
                                         "GCRMA" = "GCRMA"), 
                             selectize = T),
                 tags$hr()
)


```

```{r data reactive}
# Reactive values to store imported data
data <- reactive({
  
  # Err if NULL
  if(is.null(input$upload)){
    return(NULL)
  }
  
  # Conditionals for data source type
  if(input$dataType == "Nanostring"){
    showNotification("Processing Nanostring...", type = "message")

    # Import RCC files and convert to table/metadata
    data <- import_rcc2(input$upload$datapath)
    
    # number of genes
    geneN <- nrow(data$x)
  
    # number the samples
    sampleN <- ncol(data$x)-3
  
    # sample ID's
    sampleID <- colnames(data$header)
    
    # un-normalized data for boxplot
    unnorm_boxplot <- subset(data$x, select = c(-1,-3)) %>% 
         reshape2::melt(id.vars = c("Name")) 
    
    # Normalized data (fix later with new normalization techniques)
    if(input$normalizationNS != "deseq2"){
      normalized <- NanoStringNorm(
        x = data$x,
        anno = NA,
        CodeCount = 'sum',
        Background = 'mean',
        SampleContent = input$normalizationNS,
        round.values = FALSE,
        take.log = FALSE,
        return.matrix.of.endogenous.probes = TRUE
      )
      
      # normalized data for boxplot
      norm_boxplot <- melt(normalized, id.vars = c("Name"), varnames = c("gene", "variable")) 
      
    }
    
    # Show progress message
    showNotification("Processing complete...", type = "message")
    
    # Return a list of genes and metadata
    return(
      list(
        sampleN = sampleN,
        sampleID = sampleID,
        geneN = geneN,
        normalized = normalized,
        unnorm_boxplot = unnorm_boxplot,
        norm_boxplot = norm_boxplot)
    )
    
  }else if(input$dataType == "Microarray"){
    showNotification("Processing Microarray..", type = "message")
    
    # Load data with affy package
    ab <- ReadAffy(filenames = input$upload$datapath, compress = T, sampleNames = input$upload$name)
    
    # Un normalized data for boxplot. Temporary speed fix. first 100,000 probe values
    unnorm_boxplot <- melt(pm(ab), id.vars = c("Name"), varnames = c("gene", "variable"))[1:100000,]

    # sample N
    sampleN <- length(pData(ab)$sample)

    # SampleID
    sampleID <- row.names(pData(ab))

    # Normalize based on 
    if (input$normalizationMA == "RMA"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 20)
      normalized <- rma(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else if (input$normalizationMA == "MAS5"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 15)
      normalized <- mas5(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else if (input$normalizationMA == "GCRMA"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 15)
      normalized <- gcrma(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else{
      showNotification("Error processing data...", type = "error")
      return()
    }
    
    # Show progress message
    showNotification("Processing complete...", type = "message")
    
    # Return a list of genes and metadata
    return(
      list(
        sampleN = sampleN,
        sampleID = sampleID,
        geneN = geneN,
        normalized = normalized,
        unnorm_boxplot = unnorm_boxplot,
        norm_boxplot = norm_boxplot)
    )
    
  } # end of microarray
  
})
```


Row
-----------------------------------------------------------------------

### Number of genes

```{r valuebox geneN}
output$geneNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data())){
    return(valueBox(0, caption = "Number of genes", icon = "fa-cubes"))
  }
  valueBox(data()$geneN, caption = "Number of genes", icon = "fa-cubes")
})
valueBoxOutput(outputId = "geneNumber")
```

### Number of samples

```{r valuebox sampleN}
output$sampleNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data()$sampleN)){
    return(valueBox(0, caption = "Number of samples", icon = "fa-users"))
  }
  
  valueBox(data()$sampleN, caption = "Number of samples", icon = "fa-users")
})
valueBoxOutput(outputId = "sampleNumber")
```


Row 
-------------------------------------
   
###  

```{r raw_boxplot}
output$unnormalized <- renderHighchart({
  
  # Err if NULL
  if(is.null(data()$unnorm_boxplot)){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Plotting raw values...", type = "message")
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(data()$unnorm_boxplot$value), 
                          by = data()$unnorm_boxplot$variable,
                          color = "#2980b9")  %>%
    hc_title(text = "Raw Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("unnormalized")
``` 

###  

```{r norm_boxplot}
output$normalized <- renderHighchart({
  
  # Err if NULL
  if(is.null(data()$norm_boxplot)){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Plotting normalized values...", type = "message")
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(data()$norm_boxplot$value), 
                          by = data()$norm_boxplot$variable,
                          color = "#2980b9")  %>%
    hc_title(text = "Normalized Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("normalized")
``` 


Experimental setup
===================================== 

Inputs {.sidebar data-width=267}
-----------------------------------------------------------------------
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed commodo bibendum imperdiet. Pellentesque pharetra hendrerit feugiat. Fusce ultrices malesuada purus, quis vehicula quam. Nulla facilisi. Phasellus pulvinar lacus id ante scelerisque eleifend. 

```{r save experiment}
# Save experiemental information
actionButton(inputId = "saveExp", icon = icon("floppy-o"), label = "Save", width = "215px")

# Store all reactive data assocaited with Series information
reactive <- reactiveValues()

# On save, add values to appropriate sections
observeEvent(input$saveExp, {
  showNotification("Saving data...", type = "message")
  
  # TODO validations
  reactive$name <- isolate(input$name)
  reactive$title <- isolate(input$title)
  reactive$summary <- isolate(input$summary)
  reactive$design <- isolate(input$design)
  
  reactive$organism <- isolate(input$organism)
  reactive$molecule <- isolate(input$molecule)
  reactive$source_type <- isolate(input$source_type)
  reactive$platform <- isolate(input$platform)
  reactive$pubmed <- isolate(input$pubmed)
  
  reactive$extraction <- isolate(input$extraction)
  reactive$labeling <- isolate(input$labeling)
  reactive$hybridization <- isolate(input$hybridization)
  reactive$scanning <- isolate(input$scanning)
  reactive$data_processing <- isolate(input$data_processing)

  # Add sampleN for easier looping
  if(is.null(data())){
    reactive$sampleN <- 0
  } else{
    reactive$sampleN <- isolate(data()$sampleN)
  }
  
})

```


Column {.tabset .tabset-fade}
-------------------------------------

### Experimental 
    
```{r experimental}
h4("Name")
textInput(inputId = "name", 
          label = "Provide an identifier for this entity. This identifier is used only as an internal reference within a given file. The identifier will not appear on final GEO records.", 
          width = "400px",
          value = "", 
          placeholder = "Enter text...")

tags$hr()

h4("Title")
textInput(inputId = "title", 
          label = "Provide a unique title that describes the overall study. Max 255 characters.", 
          value = "", 
          width = "400px",
          placeholder = "Enter text...")

tags$hr()

h4("Summary")
textareaInput(inputId = "summary", 
          label = "Summarize the goals and objectives of this study. The abstract from the associated publication may be suitable. You can include as much text as you need to thoroughly describe the study.", 
          value = "", 
          placeholder = "Enter text...")

h4("Design")
textareaInput(inputId = "design", 
          label = "Provide a description of the experimental design. Indicate how many Samples are analyzed, if replicates are included, are there control and/or reference Samples", 
          value = "", 
          placeholder = "Enter text...")
```
   
### Platform 

```{r platform}
h4("Organism")
textInput(inputId = "organism", 
          label = "Identify the organism(s) from which the biological material was derived. See http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/", 
          value = "",
          width = "400px",
          placeholder = "Mus musculus")

tags$hr()

h4("Molecule Type")
selectInput(inputId = 'molecule', 
            label = 'Specify the type of molecule that was extracted from the biological material.', 
            choices = c("total RNA", "polyA RNA", "cytoplasmic RNA", "nuclear RNA", "genomic DNA", "protein", "other"),
            width = "400px",
            selectize=TRUE)

tags$hr()

h4("Sample Source")
textInput(inputId = "source_type", 
          label = "Enter sample source type.", 
          value = "", 
          width = "400px",
          placeholder = "Terminal Ileum")

tags$hr()

h4("Platform")
textInput(inputId = "platform", 
          label = "Enter the NCBI platform identifier used. See http://www.ncbi.nlm.nih.gov/geo/browse/?view=platforms&tool=findplatform to find the corresponding identifier.", 
          value = "", 
          width = "400px",
          placeholder = "GPL19964")

tags$hr()

h4("Label Type")
textInput(inputId = "label", 
          label = "Specify the compound used to label the extract e.g., biotin, Cy3, Cy5, 33P.", 
          value = "", 
          width = "400px",
          placeholder = "biotin")

tags$hr()

h4("Pubmed ID (Optional)")
textInput(inputId = "pubmed", 
          label = "Optionally, specify a valid PubMed identifier (PMID) that references a published article describing this study. Most commonly, this information is not available at the time of submission - it can be added later once the data are published.", 
          value = "", 
          width = "400px",
          placeholder = "12345678")
```     


### Protocols 

```{r protocols}
h4("Extraction Protocol")
textareaInput(inputId = "extraction", 
              label = "Describe the protocol used to isolate the extract material. You can include as much text as you need to thoroughly describe the protocol.", 
              value = "TRIzol (Invitrogen) followed by RNeasy column cleanup (Qiagen) using the manufacturers protocols", 
              placeholder = "Enter text...")

h4("Labeling Protocol")
textareaInput(inputId = "labeling", 
              label = "Describe the protocol used to label the extract. You can include as much text as you need to thoroughly describe the protocol.", 
              value = "", 
              placeholder = "Enter text...")

h4("Hybridization Protocol")
textareaInput(inputId = "hybridization", 
              label = "Describe the protocols used for hybridization, blocking and washing, and any post-processing steps such as staining.", 
              value = "", 
              placeholder = "Enter text...")

h4("Scanning Protocol")
textareaInput(inputId = "scanning", 
              label = "Describe the scanning and image acquisition protocols, hardware, and software. You can include as much text as you need to thoroughly describe the protocol.", 
              value = "", 
              placeholder = "Enter text...")

h4("Data Processing Protocol")
textareaInput(inputId = "data_processing", 
              label = "Provide details of how data in the VALUE column of your table were generated and calculated, i.e., normalization method, data selection procedures and parameters, transformation algorithm (e.g., MAS5.0), and scaling parameters. You can include as much text as you need to thoroughly describe the processing procedures.", 
              value = "", 
              placeholder = "Enter text...")
```


### Authors
    
```{r author input}
fluidRow(
    column(width = 4, 
           textInput(inputId = "firstName", 
                     label = h5("First Name"), 
                     value = "", 
                     placeholder = "Watson")),
    column(width = 4, 
           textInput(inputId = "middleInitial", 
                     label = h5("Middle Initial (optional)"), 
                     value = "", 
                     placeholder = "N")),
    column(width = 4, 
           textInput(inputId = "lastName", 
                     label = h5("Last Name"), 
                     value = "", 
                     placeholder = "Crick"),
           actionButton(inputId = "addAuthor", label = "Add Author", icon = icon("fa-list")),
           actionButton(inputId = "removeAuthor", label = "Remove Author", icon = icon("fa-list")))
    )
tags$hr()
```   


```{r author table}
reactive$authorDF <- data.frame(First = "",
                                Middle = "",
                                Last = "")

# Observe if 'add author' button is pressed
observeEvent(input$addAuthor, {
  
  # Require first and last names
  if(input$firstName == "" | input$lastName == ""){
    showNotification("Error Adding Author...", type = "error")
  }
  
  # Require first and last names
  req(input$firstName)
  req(input$lastName)
  
  # Add values to data frame (for viewing table)
  showNotification("Adding Author...", type = "message")
  author_newLine <- isolate(c(input$firstName, input$middleInitial, input$lastName))
  isolate(reactive$authorDF <- rbind(as.matrix(reactive$authorDF), unlist(author_newLine)))
  
  # Concatenate names together (for internal use)
  new_author <- isolate(
    paste(input$firstName, input$middleName, input$lastName, sep = ",") %>% 
      gsub(pattern = ",,", replacement = ",", x = ., fixed = T) %>%  # fix no middle initials
      gsub(pattern = ".", replacement = "", x = ., fixed = T) # fix dots in middle initials
  )
  
  # Add authors as a reactive list
  reactive$authorList <- c(isolate(reactive$authorList), new_author)
})

# Render table of authors
output$author_table <- renderTable({reactive$authorDF}, include.rownames = F)
tableOutput("author_table")
```


Sample data
=====================================

Inputs {.sidebar data-width=267}
-----------------------------------------------------------------------
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed commodo bibendum imperdiet. Pellentesque pharetra hendrerit feugiat. Fusce ultrices malesuada purus, quis vehicula quam. Nulla facilisi. Phasellus pulvinar lacus id ante scelerisque eleifend. 

```{r save sample}
# Save sample information
actionButton(inputId = "saveSample", icon = icon("floppy-o"), label = "Save", width = "215px")
observeEvent(input$saveSample, {
  showNotification("Saving data...", type = "message")
})
```


Column {data-width=300}
-------------------------------------
    
### Selection
    
```{r sample characteristic}
h4("Choose characteristics")
selectizeInput(inputId = "sampleChar", 
               label = helpText("Fusce ultrices malesuada purus, quis vehicula quam. Nulla facilisi. Phasellus pulvinar lacus id ante scelerisque eleifend."), 
            choices = c("dose", "time", "tissue", "strain", "gender", "cell line", "development stage", "age", "agent", "cell type", "infection", "isolate", "metabolism", "shock", "stress", "temperature", "specimen", "disease state", "protocol", "growth protocol", "genotype/variation", "species", "individual"), 
            options = list(create = TRUE, placeholder = 'Select variables...'),
            multiple = TRUE)
```


Column 
-------------------------------------
    
### Characteristics Table
    
```{r sample select}
output$hot = renderRHandsontable({
  
  # Err if null
  if(is.null(data())){
    return()
  }
  
  # Get number of sample
  N <- as.integer(data()$sampleN)
  
  # Create new table with row.names as sampe id's
  df <- data.frame(ID = row.names(input$upload), 
                   Sample = data()$sampleID, 
                   Description = character(N))
  
  # aggregate columns
  cols <- input$sampleChar
  
  # Dynamticlaly add new columns to df
  for(i in cols){
    df[i] <- character(N)
  }
  
  # Generate handsome table
  rht <- rhandsontable(df, stretchH = "all", rowHeaders = NULL) %>%
    hot_col("Sample", readOnly = TRUE, copyable = T) %>%
    hot_col("ID", readOnly = TRUE, copyable = T) %>%
    hot_col("Description", type = "autocomplete", copyable = T) %>%
    hot_context_menu(allowRowEdit = FALSE, allowColEdit = FALSE)
  
  # Make all colums factorable
  for(i in cols){
    rht <- rht %>%
      hot_col(i, "autocomplete", copyable = T, highlightRow = F, strict = T)
  }
  
  # Return table
  rht
  
})

# Output Hot table
fillCol(rHandsontableOutput("hot"))
```



Review & Export
=====================================

Inputs {.sidebar data-width=267}
-----------------------------------------------------------------------
The tool will export a Simple Omnibus Format in Text (SOFT) file, which will include all relavent information for a Minimum Information About a Microarray Experiment (MIAME) compliant experiment.

```{r}
downloadLink(outputId = 'downloadData', label = 'Download')
```

Row
-------------------------------------
    
### Author
    
```{r}
output$test1 <- renderPrint({ 
  reactive$name
})
verbatimTextOutput('test1')
```
 
### Chart 2
    
```{r}
output$test2 <- renderPrint({ 
  if(is.null(input$hot)){
   return("invalid")
  }
  hot_to_r(input$hot)
})
verbatimTextOutput('test2')
``` 

Row
-------------------------------------
    
### Chart 3
    
```{r}
output$test3 <- renderPrint({ 
  if(is.null(data())){
    return()
  }
  c(data()$sampleN,
    data()$sampleID,
    data()$geneN)
        #data()$normalized,
        #data()$unnorm_boxplot,
        #data()$norm_boxplot)
})
verbatimTextOutput('test3')
```
    
### Export

```{r}
output$downloadData <- downloadHandler(
  filename = function() { 
    paste("soft", '.txt', sep = '') 
  },
  
  content = function(file) {
    
    # Write series information
    write_append(paste0("^SERIES = ", isolate(reactive$name)), filename = file)
    write_append(paste0("!Series_title = ", isolate(reactive$title)), filename = file)
    write_append(paste0("!Series_summary = ", isolate(reactive$summary)), filename = file)
    write_append(paste0("!Series_overall_design = ", isolate(reactive$design)), filename = file)

    # Loop over each number of samples
    for(i in 1:isolate(reactive$sampleN)){
      write_append(paste0("!Series_sample_id = ", as.character(i)), filename = file)
    }
    
    # Loop over each of the authors
    for(i in isolate(reactive$authorList)){
      write_append(paste0("!Series_contributor = ", i), filename = file)
    }
    
    # Sample information
    sampleInfo <- hot_to_r(input$hot)
  # sampleInfo <- sampleInfo[!is.null(sampleInfo$SampleID), ] # Temporary fix to remove NA rows (Issue #2)
    for (i in 1:nrow(sampleInfo)){
      
      # Write ^SAMPLE
      write_append(paste0("^SAMPLE", " = ", sampleInfo$ID[i]), filename = file)
      
      # Write !Sample_title
      write_append(paste0("!Sample_title", " = ", sampleInfo$Sample[i]), filename = file)
      
      # Write !Sample_supplementary_file
      write_append(paste0("!Sample_supplementary_file", " = ", input$upload$name[i]), filename = file)
      
      # Write !Sample_supplementary_file
      sub <- sampleInfo[i, c(-1:-3)]
      sapply(colnames(sub), function(x) {
         write_append(paste0("!Sample_characteristics_ch1", " = ", x, " : ", sub[ ,x]), filename = file)
      })
      
      # Write !Sample_Description
      write_append(paste0("!Sample_Description", " = ", sampleInfo$Description[i]), filename = file)
      
      # Write !Sample_source_name
      write_append(paste0("!Sample_source_name", " = ", isolate(reactive$source_type)), filename = file)
      
      # Write !Sample_organism
      write_append(paste0("!Sample_organism", " = ", isolate(reactive$organism)), filename = file)

      # Write !Sample_label
      write_append(paste0("!Sample_label", " = ", isolate(reactive$label)), filename = file)
      
      # Write !Sample_molecule
      write_append(paste0("!Sample_molecule", " = ", isolate(reactive$molecule)), filename = file)
      
      # Write !Sample_extract_protocol
      write_append(paste0("!Sample_extract_protocol", " = ", isolate(reactive$extraction)), filename = file)
      
      # Write !Sample_label_protocol
      write_append(paste0("!Sample_label_protocol", " = ", isolate(reactive$labeling)), filename = file)
      
      # Write !Sample_hyb_protocol
      write_append(paste0("!Sample_hyb_protocol", " = ", isolate(reactive$hybridization)), filename = file)
      
      # Write !Sample_scan_protocol
      write_append(paste0("!Sample_scan_protocol", " = ", isolate(reactive$scanning)), filename = file)
      
      # Write !Sample_data_processing
      write_append(paste0("!Sample_data_processing", " = ", isolate(reactive$data_processing)), filename = file)
      
      # Subset the expression values
      nsn_sub <- subset(data()$normalized, select = i)
      
      # Iterate over each row and write the expression values
      write_append("#ID_REF =", filename = file)
      write_append(paste0("#VALUE = ", "Normalized expression counts"), filename = file)
      write_append("!Sample_table_begin", filename = file)
      write_append(paste("ID_REF", "VALUE", sep = "\t"), filename = file)
      write.table(as.matrix(nsn_sub), 
                  file = file, 
                  append = TRUE, 
                  quote = FALSE, 
                  row.names = TRUE, 
                  col.names = FALSE, 
                  sep = "\t")
      write_append("!Sample_table_end", filename = file)
    }
})
```



<style>
#sidebar.section.sidebar {
  background-color: white; 
  font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif !important;
}

.js-irs-0 .irs-bar {
border-top-color: #d01010;
border-bottom-color: #d01010;
} 

.js-irs-0 .irs-bar-edge {
border-color: #d01010;
}

.js-irs-0 .irs-single, .js-irs-0 .irs-bar-edge, .js-irs-0 .irs-bar {
background: #a00;
}

</style>
