---
title: "GEO-submitter"
runtime: shiny
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
    source_code: embed
    theme: bootstrap
---

```{r setup, include=FALSE}
library(flexdashboard)
library(shiny)
source("util/helpers.R")
source("util/import_rcc.R")
options(shiny.maxRequestSize = 9*1024^2) # 9mb limit
tags$head(tags$script(src = "www/js/google-analytics.js")) # google analystics
```


Introduction {data-orientation=columns}
=====================================  


Column {data-width=700}
-------------------------------------
    
### Introduction {data-height=250}
    
```{r introduction}
p('Currently there does not exist an easy-to-use interface for submitting array data to Gene Expression Omnibus (GEO), through NCBI. This difficulty can lead to the inability to properly submit studes and adhere to the', a('Minimum information about a microarray experiment (MAIME).', href = "http://www.ncbi.nlm.nih.gov/geo/info/MIAME.html", target="_blank"), 'GEO-submiter is meant to remedy this issue by creating a web-based tool to help researchers submit their gene expression array studies to GEO by using the, ', a("Simple Omnibus Format in Text (SOFT).", href = 'http://www.ncbi.nlm.nih.gov/geo/info/soft.html', target="_blank"), 'The SOFT file format allows a researcher to include all experiemental protocols, information and gene expression data in one file. After filling in all required information found within this tool, a SOFT text file will be generated, which can be submitted along with the raw data to GEO as a single submission.' )

tags$b("This is still a work-in-progress and a proof-of-concept, so do not use it for anything important!")

```

### Normalization methods 
    
```{r normalization methods}
p('GEO submission require the SOFT file to include normalized expression values for each of the genes for every sample within an array. GEO-submitter makes it easier to generate the normalized values using a multitude of R-packages. After importing the raw data, you can select the many different techniques for normalizing your microarray or nanostring data. Below is a breakdown of the different normalization techniques currently supported by GEO-submitter.')

h4("Microarray Normalization")
p("Robust Multi-array Average (RMA)", a("Paper", href = "http://www.ncbi.nlm.nih.gov/pubmed/12925520", target="_blank"))
p("(MAS5)", a("Paper", href = "http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-8-273", target="_blank"))
p("(GCRMA)", a("Bioconductor", href = "https://www.bioconductor.org/packages/release/bioc/html/gcrma.html", target="_blank"))

tags$hr()

h4("Nanostring Normalization")
p("NanoStringNorm-Sum (housekeeping.sum)")
p("NanoStringNorm-Mean (housekeeping.geo.mean)")
```


Column {data-width=300}
-------------------------------------

    
### Credits & citations

```{r credits}
p("Below are a list of the packages and databases which were used or influenced the creation GEO-submitter. See the source code for a more detailed listing of the packages used.")
tags$hr()
h5("Shiny:", a("Github, ", href = 'https://github.com/rstudio/shiny', target="_blank"), a("Website", href = 'http://shiny.rstudio.com/', target="_blank"))
tags$hr()
h5("Flexdashboard:", a("Github, ", href = 'https://github.com/rstudio/flexdashboard', target="_blank"), a("Website", href = 'http://rmarkdown.rstudio.com/flexdashboard/index.html', target="_blank"))
tags$hr()
h5("NanoStringNorm:", a("Paper, ", href = 'http://www.ncbi.nlm.nih.gov/pubmed/22513995', target="_blank"), a("CRAN", href = 'https://cran.r-project.org/web/packages/NanoStringNorm/index.html', target="_blank"))
tags$hr()
h5("GEOsubmission:", a("Bioconductor", href = 'http://www.bioconductor.org/packages//release/bioc/html/GEOsubmission.html', target="_blank"))
tags$hr()
h5("highcharter:", a("Github", href = 'http://www.ncbi.nlm.nih.gov/pubmed/11726920', target="_blank"))
tags$hr()
h5("rhandsontable:", a("Github", href = 'https://github.com/jrowen/rhandsontable', target="_blank"))
tags$hr()
h5("affy:", a("Bioconductor", href = 'https://bioconductor.org/packages/release/bioc/html/affy.html', target="_blank"))
tags$hr()
h5("dplyr:", a("Github", href = 'https://github.com/hadley/dplyr', target="_blank"), a("Website", href = 'https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html', target="_blank"))
tags$hr()
h5("GEO:", a("Website", href = 'http://www.ncbi.nlm.nih.gov/geo/', target="_blank"))
tags$hr()
h5("MAIME:", a("Paper", href = 'http://www.ncbi.nlm.nih.gov/pubmed/11726920', target="_blank"))
tags$hr()
h5("Bootstrap3:", a("Github", href = 'https://github.com/twbs/bootstrap', target="_blank"))
tags$hr()
h5('Application author', a('Thomas W. Battaglia', href = 'https://github.com/twbattaglia', target="_blank"))

```



Upload {data-orientation=rows}
=====================================  

Inputs {.sidebar data-width=268}
-----------------------------------------------------------------------

```{r upload bar}
# Data Type
h4('Data Type')
selectizeInput(inputId = 'dataType', width = NULL,
               label = helpText("Select an array platform type."), 
               choices = c('Nanostring', "Microarray"), 
               options = list(placeholder = 'Select array type....     ', 
                              onInitialize = I('function() { this.setValue(""); }'))
               )
tags$hr()

# Input Files
h4('Upload Files')
fileInput(inputId = 'upload', 
          label = helpText("Use 'Shift' to select multiple files."), 
          multiple = T,
          accept = c(
            '.RCC',
            '.RCC.gz', 
            '.CEL',
            '.CEL.gz')
      )
tags$hr()

# Normalization method
conditionalPanel(condition = "input.dataType == 'Nanostring'",
                 h4('Normalization Method'),
                 selectInput(inputId = "normalizationNS", 
                             label = helpText("Select a normalization method"), 
                             choices = c("Housekeeping-Sum" = "housekeeping.sum",
                                         "Housekeeping-Mean" = "housekeeping.geo.mean"), 
                             selectize = T),
                 tags$hr()
)

conditionalPanel(condition = "input.dataType == 'Microarray'",
                 h4('Normalization Method'),
                 selectInput(inputId = "normalizationMA", 
                             label = helpText("Select normalization method."), 
                             choices = c("RMA" = "RMA",
                                         "MAS5" = "MAS5", 
                                         "GCRMA" = "GCRMA"), 
                             selectize = T),
                 tags$hr()
)
```

```{r data reactive}
# Reactive values to store imported data
data <- reactive({
  
  # Err if NULL
  if(is.null(input$upload)){
    return(NULL)
  }
  
  # Conditionals for data source type
  if(input$dataType == "Nanostring"){
    showNotification("Processing Nanostring...", type = "message")

    # Import RCC files and convert to table/metadata
    data <- import_rcc2(input$upload$datapath)
    
    # number of genes
    geneN <- nrow(data$x)
  
    # number the samples
    sampleN <- ncol(data$x)-3
  
    # sample ID's
    sampleID <- colnames(data$header)
    
    # un-normalized data for boxplot
    unnorm_boxplot <- subset(data$x, select = c(-1,-3)) %>% 
         reshape2::melt(id.vars = c("Name"), varnames = c("gene", "variable")) 
    
    # Normalized data (fix later with new normalization techniques)
    if(input$normalizationNS != "deseq2"){
      normalized <- NanoStringNorm(
        x = data$x,
        anno = NA,
        CodeCount = 'sum',
        Background = 'mean',
        SampleContent = input$normalizationNS,
        round.values = FALSE,
        take.log = FALSE,
        return.matrix.of.endogenous.probes = TRUE
      )
      
      # normalized data for boxplot
      norm_boxplot <- melt(normalized, id.vars = c("Name"), varnames = c("gene", "variable")) 
    }
    
    # Show progress message
    showNotification("Processing complete...", type = "message")
    
    # Return a list of genes and metadata
    return(
      list(
        sampleN = sampleN,
        sampleID = sampleID,
        geneN = geneN,
        normalized = normalized,
        unnorm_boxplot = unnorm_boxplot,
        norm_boxplot = norm_boxplot)
    )
    
  }else if(input$dataType == "Microarray"){
    showNotification("Processing Microarray..", type = "message")
    
    # Load data with affy package
    ab <- ReadAffy(filenames = input$upload$datapath, compress = T, sampleNames = input$upload$name)
    
    # Un normalized data for boxplot. Temporary speed fix. sample 10,000 from each sample
    unnorm_boxplot <- melt(pm(ab), varnames = c("gene", "variable")) %>%
      group_by(variable) %>%
      sample_n(10000)
    
    # sample N
    sampleN <- length(pData(ab)$sample)

    # SampleID
    sampleID <- row.names(pData(ab))

    # Normalize based on 
    if (input$normalizationMA == "RMA"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 20)
      normalized <- rma(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else if (input$normalizationMA == "MAS5"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 15)
      normalized <- mas5(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else if (input$normalizationMA == "GCRMA"){
      showNotification("Normalizing Data. Please wait...", type = "warning", duration = 15)
      normalized <- gcrma(ab)
      geneN <- nrow(normalized@featureData@data)
      norm_boxplot <- melt(exprs(normalized), id.vars = c("Name"), varnames = c("gene", "variable")) 

    } else{
      showNotification("Error processing data...", type = "error")
      return()
    }
    
    # Show progress message
    showNotification("Processing complete...", type = "message")
    
    # Return a list of genes and metadata
    return(
      list(
        sampleN = sampleN,
        sampleID = sampleID,
        geneN = geneN,
        normalized = exprs(normalized),
        unnorm_boxplot = unnorm_boxplot,
        norm_boxplot = norm_boxplot)
    )
    
  } # end of microarray
  
})
```


Row
-----------------------------------------------------------------------

### Number of genes

```{r valuebox geneN}
output$geneNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data())){
    return(valueBox(0, caption = "Number of genes", icon = "fa-cubes"))
  }
  valueBox(data()$geneN, caption = "Number of genes", icon = "fa-cubes")
})
valueBoxOutput(outputId = "geneNumber")
```

### Number of samples

```{r valuebox sampleN}
output$sampleNumber <- renderValueBox({
  
  # Err if NULL
  if(is.null(data()$sampleN)){
    return(valueBox(0, caption = "Number of samples", icon = "fa-users"))
  }
  
  valueBox(data()$sampleN, caption = "Number of samples", icon = "fa-users")
})
valueBoxOutput(outputId = "sampleNumber")
```


Row 
-------------------------------------
   
###  

```{r raw_boxplot}
output$unnormalized <- renderHighchart({
  
  # Err if NULL
  if(is.null(data()$unnorm_boxplot)){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Plotting raw values...", type = "message")
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(data()$unnorm_boxplot$value), 
                          by = data()$unnorm_boxplot$variable,
                          color = "#2980b9")  %>%
    hc_title(text = "Raw Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("unnormalized")
``` 

###  

```{r norm_boxplot}
output$normalized <- renderHighchart({
  
  # Err if NULL
  if(is.null(data()$norm_boxplot)){
    return(NULL)
  }
  
  # Convert to long format for high charts
  showNotification("Plotting normalized values...", type = "message")
  
  # Plot boxplot over gene abundance values
  highchart() %>% 
    hc_add_series_boxplot(x = log(data()$norm_boxplot$value), 
                          by = data()$norm_boxplot$variable,
                          color = "#2980b9")  %>%
    hc_title(text = "Normalized Expression Values") %>%
    hc_legend(enabled = FALSE) %>%
    hc_yAxis(title = list(text = "Log10 gene abundances"))
})
highchartOutput("normalized")
``` 


Experimental setup
===================================== 

Inputs {.sidebar data-width=268}
-----------------------------------------------------------------------

```{r save experiment}
h4("Help")
p("This section is the heart of", tags$b('GEO-submitter '), "and should be filled completely and as detailed as possible. The information stored here is meant for researchers to understand the experimental design, as well as the platform and protocols used throughout the study.")
  
p("The best way to write detailed protocols and experimental metadata, is to view studies similar to your own. Search", a("GEO DataSets", href = "http://www.ncbi.nlm.nih.gov/gds", target="_blank"), " for submitted and curated studies. It is also useful to query studies with the same platform identifer, so you can re-use the protocols for labeling, hybridization, etc..")

p("When you have filled out the entries completly, save the information using the button below.")

# Save experiemental information
actionButton(inputId = "saveExp", icon = icon("floppy-o"), label = "Save", width = "215px")

# Store all reactive data assocaited with Series information
reactive <- reactiveValues()

# On save, add values to appropriate sections
observeEvent(input$saveExp, {
  showNotification("Saving data...", type = "message")
  
  # TODO validations
  reactive$Name <- isolate(input$name)
  reactive$Title <- isolate(input$title)
  reactive$Summary <- isolate(input$summary)
  reactive$Design <- isolate(input$design)
  
  reactive$Organism <- isolate(input$organism)
  reactive$Molecule <- isolate(input$molecule)
  reactive$Source_type <- isolate(input$source_type)
  reactive$Platform <- isolate(input$platform)
  reactive$Pubmed <- isolate(input$pubmed)
  reactive$Label <- isolate(input$label)

  reactive$Extraction_protocol <- isolate(input$extraction)
  reactive$Labeling_protocol <- isolate(input$labeling)
  reactive$Hybridization_protocol <- isolate(input$hybridization)
  reactive$Scanning_protocol <- isolate(input$scanning)
  reactive$Data_protocol <- isolate(input$data_processing)
  
  reactive$Biomaterial_protocol <- isolate(input$biomaterial)
  reactive$Treatment_protocol <- isolate(input$treatment)
  reactive$Growth_protocol <- isolate(input$growth)

  # Add sampleN for easier looping
  if(is.null(data())){
    reactive$sampleN <- 0
  } else{
    reactive$sampleN <- isolate(data()$sampleN)
  }
  
})

```


Column {.tabset .tabset-fade}
-------------------------------------

### Experimental 
    
```{r experimental}
h4("Name")
p("Provide an identifier for this entity. This identifier is used only as an internal reference within a given file. The identifier will not appear on final GEO records.")
textInput(inputId = "name", 
          label = NULL, 
          width = "375px",
          value = "", 
          placeholder = "Bone_marrow_stromal_cells")

tags$hr()

h4("Title")
p("Provide a unique title that describes the overall study. Max 255 characters.")
textAreaInput(inputId = "title", 
          label = NULL, 
          value = "", 
          rows = 2,
          width = "750px",
          placeholder = "Profiling of the functionally distinct human bone marrow stromal cell lines HS-5 and HS-27a.")

tags$hr()

h4("Summary")
p("Summarize the goals and objectives of this study. The abstract from the associated publication may be suitable. You can include as much text as you need to thoroughly describe the study.")
textAreaInput(inputId = "summary", 
              label = NULL, 
              value = "", 
              rows = 5, 
              resize = 'vertical',
              width = "750px",
              placeholder = "Enter text...")

tags$hr()

h4("Design")
p("Provide a description of the experimental design. Indicate how many Samples are analyzed, if replicates are included, are there control and/or reference samples. etc..")
textAreaInput(inputId = "design", 
              label = NULL, 
              value = "", 
              rows = 3, 
              resize = 'vertical',
              width = "750px", 
              placeholder = "Two-condition experiment, ES vs. TF-ES cells. Biological replicates: 4 control, 3 transfected, independently grown and harvested. One replicate per array.")
```
   
### Platform 

```{r platform}
h4("Organism")
p("Identify the organism(s) from which the biological material was derived. See", a("NCBI Taxonomy", href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/", target="_blank"), " for a listing.")
textInput(inputId = "organism", 
          label = NULL,
          value = "",
          width = "375px",
          placeholder = "Mus musculus")

tags$hr()

h4("Molecule Type")
p('Specify the type of molecule that was extracted from the biological material.')
selectInput(inputId = 'molecule', 
            label = NULL, 
            choices = c("total RNA", 
                        "polyA RNA", 
                        "cytoplasmic RNA", 
                        "nuclear RNA", 
                        "genomic DNA", 
                        "protein", 
                        "other"),
            width = "375px",
            selectize = TRUE)

tags$hr()

h4("Sample Source")
p("Select the sample source type.")
textInput(inputId = "source_type", 
          label = NULL, 
          value = "", 
          width = "375px",
          placeholder = "Terminal Ileum")

tags$hr()

h4("Platform")
p("Enter the NCBI platform identifier used.", a("See", href = "http://www.ncbi.nlm.nih.gov/geo/browse/?view=platforms&tool=findplatform", target="_blank"), "to find the corresponding identifier.")
textInput(inputId = "platform", 
          label = NULL, 
          value = "", 
          width = "375px",
          placeholder = "GPL19964")

tags$hr()

h4("Label Type")
p("Specify the compound used to label the extract e.g., biotin, Cy3, Cy5, 33P.")
textInput(inputId = "label", 
          label = NULL, 
          value = "", 
          width = "375px",
          placeholder = "biotin")

tags$hr()

h4("Pubmed ID (Optional)")
p("Optionally, specify a valid PubMed identifier (PMID) that references a published article describing this study. Most commonly, this information is not available at the time of submission - it can be added later once the data are published.")
textInput(inputId = "pubmed", 
          label = NULL, 
          value = "", 
          width = "375px",
          placeholder = "12345678")
```     


### Protocols 

```{r protocols}
h4("Extraction Protocol")
p("Describe the protocol used to isolate the extract material. You can include as much text as you need to thoroughly describe the protocol.")
textAreaInput(inputId = "extraction", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px",
              placeholder = "TRIzol (Invitrogen) followed by RNeasy column cleanup (Qiagen) using the manufacturers’ protocols")

tags$hr()

h4("Labeling Protocol")
p("Describe the protocol used to label the extract. You can include as much text as you need to thoroughly describe the protocol.")
textAreaInput(inputId = "labeling", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px",
              placeholder = "Affymetrix GeneChip® IVT Labeling Kit, according to manufactuer's protocol (GeneChip® Expression Analysis Technical Manual)")

tags$hr()

h4("Hybridization Protocol")
p("Describe the protocols used for hybridization, blocking and washing, and any post-processing steps such as staining.")
textAreaInput(inputId = "hybridization", 
              label = NULL,
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px",
              placeholder = "Affymetrix Eukaryotic Target Hybridization protocol (GeneChip® Expression Analysis Technical Manual)")

tags$hr()

h4("Scanning Protocol")
p("Describe the scanning and image acquisition protocols, hardware, and software. You can include as much text as you need to thoroughly describe the protocol.")
textAreaInput(inputId = "scanning", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px",
              placeholder = "Affymetrix® GeneChip® Scanner 3000 with GCOS software according to manufactuer's protocol (GeneChip® Expression Analysis Technical Manual)")

tags$hr()

h4("Data Processing Protocol")
p("Provide details of how data in the VALUE column of your table were generated and calculated, i.e., normalization method, data selection procedures and parameters, transformation algorithm (e.g., MAS5.0), and scaling parameters. You can include as much text as you need to thoroughly describe the processing procedures.")
textAreaInput(inputId = "data_processing", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px",
              placeholder = "Probe set summarization and normalization was performed by robust multi-array averaging (RMA, including background subtraction, summarization by median polish and quantile normalization) using the R/Bioconductor package affy.")

tags$hr()

h4("Biomaterial Provider (Optional)")
p("Specify the name of the company, laboratory or person that provided the biological material.")
textAreaInput(inputId = "biomaterial", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px", 
              placeholder = "Optional")
tags$hr()

h4("Growth Protocol (Optional)")
p("Describe the conditions that were used to grow or maintain organisms or cells prior to extract preparation. You can include as much text as you need to thoroughly describe the protocol.")
textAreaInput(inputId = "growth", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px", 
              placeholder = "Optional")
tags$hr()

h4("Treatment Protocol (Optional)")
p("Describe any treatments applied to the biological material prior to extract preparation. You can include as much text as you need to thoroughly describe the protocol.")
textAreaInput(inputId = "treatment", 
              label = NULL, 
              value = "", 
              rows = 4, 
              resize = 'vertical',
              width = "750px", 
              placeholder = "Optional")
tags$hr()
```


### Authors
    
```{r author input}
fluidRow(
    column(width = 4, 
           textInput(inputId = "firstName", 
                     label = h5("First Name"), 
                     value = "", 
                     placeholder = "Watson")),
    column(width = 4, 
           textInput(inputId = "middleInitial", 
                     label = h5("Middle Initial (optional)"), 
                     value = "", 
                     placeholder = "N")),
    column(width = 4, 
           textInput(inputId = "lastName", 
                     label = h5("Last Name"), 
                     value = "", 
                     placeholder = "Crick"),
           actionButton(inputId = "addAuthor", label = "Add Author", icon = icon("fa-list"))
           #actionButton(inputId = "editAuthor", label = "Edit Author List", icon = icon("fa-list"))
           )
    )
tags$hr()
```   


```{r author reactive}
# Set a reactive dataframe to show a table of authors
reactive$authorDF <- data.frame(First = character(),
                                Middle = character(),
                                Last = character())

# Observe if 'add author' button is pressed
observeEvent(input$addAuthor, {
  
  # Require first and last names
  if(input$firstName == "" | input$lastName == ""){
    showNotification("Error Adding Author...", type = "error")
  }
  
  # Require first and last names
  req(input$firstName)
  req(input$lastName)
  
  # Concatenate names together (for internal use)
  new_author <- isolate(
    paste(input$firstName, input$middleInitial, input$lastName, sep = ",") %>% 
      gsub(pattern = ",,", replacement = ",", x = ., fixed = T) %>%  # fix no middle initials
      gsub(pattern = ".", replacement = "", x = ., fixed = T) # fix dots in middle initials
  )
  
  # Add authors as a reactive list
  reactive$authorList <- c(isolate(reactive$authorList), new_author)
  
  # Add values to table
  showNotification("Adding Author...", type = "message")
  author_newLine <- isolate(c(input$firstName, input$middleInitial, input$lastName))
  isolate(reactive$authorDF <- rbind(as.matrix(reactive$authorDF), unlist(author_newLine)))
})

# Render table of authors
output$author_table <- renderTable({reactive$authorDF}, include.rownames = F)
tableOutput("author_table")
```


```{r author modal}
observeEvent(input$editAuthor, {
  showModal(modalDialog(
    title = "Edit Authors",
    h4("Select the authors you wish to remove. (TODO)"),
    renderUI({
      lapply(reactive$authorList, function(x){
        id <- gsub(pattern = ",", replacement = "_", x = x, fixed = T)
        checkboxInput(inputId = id, label = tags$b(x), value = FALSE)
      })
    }),
    actionButton(inputId = "authorRemove", label = "Remove Authors", icon = icon("fa-trash")),
    easyClose = TRUE,
    footer = NULL
  ))
})

# Remove any authors in the checklist
observeEvent(input$authorRemove, {
  
  # Iterate over each input. checkbox label will be in input$AUTHOR
  sapply(reactive$authorList, function(x){
    
    # Convert to same format as input checkbox labels so it can be found.
    id <- gsub(pattern = ",", replacement = "_", x = x, fixed = T)
    
    # If checked boxes are (TRUE), then remove them from the list of authors.
    if(input[[id]]){
      
      # Remove any entries from the author dataframe
      # - split the entry into pieces to match to the dataframe and subset based on those pieces
      split <- strsplit(x, ",", fixed = T)
      
      # Fix for if middle initial is missing
      if(length(split[[1]]) == 2){
        split[[1]][3] = split[[1]][2]
        split[[1]][2] = ""
      } 
      
      # Subset based on first, middle and last names that match
      authorTable <- as.data.frame(reactive$authorDF)
      authorTableSub <- subset(authorTable, First != as.character(split[[1]][1]) & Middle != as.character(split[[1]][3]) & Last != as.character(split[[1]][3]))
      isolate(reactive$authorDF <- authorTableSub)


      # Remove any entries from the reactive authors list
      reactive$authorList <- isolate(reactive$authorList[!reactive$authorList %in% x])
      #message(paste0("removed ", x, " from reactive list."))
    }
  })
  
})

```



Sample data
=====================================

Inputs {.sidebar data-width=268}
-----------------------------------------------------------------------

```{r save sample}
h4("Help")
p("This section is meant to give important information about each sample. Choose multiple inputs from the selection box, or type out you own if it is not listed. If information is not available for a particular sample, enter ", tags$b('Not applicable. '))
p("When you have filled out the entries completly, save the information using the button below.")


# Save sample information
actionButton(inputId = "saveSample", icon = icon("floppy-o"), label = "Save", width = "215px")
observeEvent(input$saveSample, {
  showNotification("Saving data...", type = "message")
})
```


Column {data-width=300}
-------------------------------------
    
### Selection
    
```{r sample characteristic}
h4("Choose characteristics")
selectizeInput(inputId = "sampleChar", 
               label = helpText("Select or enter the variables which best describe your samples."), 
               choices = c("dose", "time", "tissue", "strain", "gender", "cell line", "development stage", "age", "agent", "cell type", "infection", "isolate", "metabolism", "shock", "stress", "temperature", "specimen", "disease state", "protocol", "growth protocol", "genotype/variation", "species", "individual"), 
               options = list(create = TRUE, placeholder = 'Select variables...     '),
               multiple = TRUE)
```


Column 
-------------------------------------
    
### Characteristics Table
    
```{r sample select}
output$hot = renderRHandsontable({
  
  # Err if null
  if(is.null(data())){
    return()
  }
  
  # Get number of sample
  N <- as.integer(data()$sampleN)
  
  # Create new table with row.names as sampe id's
  df <- data.frame(ID = row.names(input$upload), 
                   Sample = data()$sampleID, 
                   Description = character(N))
  
  # aggregate columns
  cols <- input$sampleChar
  
  # Dynamticlaly add new columns to df
  for(i in cols){
    df[i] <- character(N)
  }
  
  # Generate handsome table
  rht <- rhandsontable(df, stretchH = "all", rowHeaders = NULL) %>%
    hot_col("Sample", readOnly = TRUE, copyable = T) %>%
    hot_col("ID", readOnly = TRUE, copyable = T) %>%
    hot_col("Description", type = "autocomplete", copyable = T) %>%
    hot_context_menu(allowRowEdit = FALSE, allowColEdit = FALSE)
  
  # Make all colums factorable
  for(i in cols){
    rht <- rht %>%
      hot_col(i, "autocomplete", copyable = T, highlightRow = F, strict = T, default = "Not applicable")
  }
  
  # Return table
  rht
  
})

# Output Hot table
fillCol(rHandsontableOutput("hot"))
```



Review & Export
=====================================

Inputs {.sidebar data-width=267}
-----------------------------------------------------------------------
```{r help export}
h4("Help")
p("Review the", tags$b("Experimental data"), "and", tags$b("Sample data"), "tables to be sure the information you have inputed is correct. Once the experimental and sample metadata are valid,", tags$b("export"), "the data as a SOFT file below.")

tags$hr()

h4("Export Data")
downloadLink('downloadData', 'Export SOFT file')

```

   
Column {.tabset .tabset-fade data-width=600}
-------------------------------------
   
### Experimental data

```{r experiment review}
output$expReview <- renderTable(hover = T, 
                                rownames = F,
                                bordered = T, {
  # Err if null                                
  if(is.null(reactive)){
    return()
  }
                                  
  x <- reactiveValuesToList(reactive)
  x$authorDF <- NULL
  x$authorList <- NULL
  df <- t(as.data.frame(x))
  data.frame(Variable = row.names(df), Values = df[,1])
})
tableOutput('expReview')
```   
 
### Sample data
```{r sample review}
output$sampleReview <- renderTable(
  hover = T, 
  bordered = T,
  rownames = F, {
  
    # Err if null
    if(is.null(data())){
      return()
    }
    df <- hot_to_r(input$hot)
    na.omit(df)
  })
tableOutput('sampleReview')
```


Column {data-width=400}
-------------------------------------
    
### Prepare and submit all data to GEO 
```{r}
p("After sucessfully generating a SOFT file, the next step is to prepare the raw data and SOFT file for GEO submission. Once complete, go to the", a("GEO-SOFT submission page", href = "http://www.ncbi.nlm.nih.gov/geo/submission/depslip.cgi", target="_blank"), " and choose the", tags$b("Submit"), " button to start a new submission.")

h5("Step 1. Place raw files and SOFT file into a single folder")
img(src = "www/folder_setup.png", width = NULL)

tags$hr()

h5("Step 2. Compress folder with raw data")
img(src = "www/compress_folder.png", width = NULL)

tags$hr()

h5("Step 3. Final folder")
img(src = "www/final_folder.png", width = NULL)

tags$hr()

h5("Step 4. Submit data")
p("Go to", a("GEO-SOFT submission page", href = "http://www.ncbi.nlm.nih.gov/geo/submission/depslip.cgi", target="_blank"))
img(src = "www/placeholder.png", width = NULL)

```



```{r export}
output$downloadData <- downloadHandler(
  contentType = "text/txt",
  filename = "soft.txt",
  content = function(file) {
    
    # Write series information
    write_append(paste0("^SERIES = ", isolate(reactive$Name)), filename = file)
    write_append(paste0("!Series_title = ", isolate(reactive$Title)), filename = file)
    write_append(paste0("!Series_summary = ", isolate(reactive$Summary)), filename = file)
    write_append(paste0("!Series_overall_design = ", isolate(reactive$Design)), filename = file)

    # Loop over each number of samples
    for(i in 1:isolate(reactive$sampleN)){
      write_append(paste0("!Series_sample_id = ", as.character(i)), filename = file)
    }
    
    # Loop over each of the authors
    for(i in isolate(reactive$authorList)){
      write_append(paste0("!Series_contributor = ", i), filename = file)
    }
    
    # Write Optionals
    if(isolate(reactive$Pubmed) != ""){
      write_append(paste0("!Series_pubmed_id", " = ", isolate(reactive$Pubmed)), filename = file)
    }
    
    # Sample information
    sampleInfo <- hot_to_r(input$hot)
    
    # Iterate over each sample
    for (i in 1:nrow(sampleInfo)){
      
      # Write ^SAMPLE
      write_append(paste0("^SAMPLE", " = ", sampleInfo$ID[i]), filename = file)
      
      # Write !Sample_title
      write_append(paste0("!Sample_title", " = ", sampleInfo$Sample[i]), filename = file)
      
      # Write !Sample_supplementary_file
      write_append(paste0("!Sample_supplementary_file", " = ", input$upload$name[i]), filename = file)
      
      # Write !Sample_supplementary_file
      sub <- sampleInfo[i, c(-1:-3)]
      sapply(colnames(sub), function(x) {
         write_append(paste0("!Sample_characteristics_ch1", " = ", x, " : ", sub[ ,x]), filename = file)
      })
      
      # Write !Sample_Description
      write_append(paste0("!Sample_Description", " = ", sampleInfo$Description[i]), filename = file)
      
      # Write !Sample_source_name
      write_append(paste0("!Sample_source_name", " = ", isolate(reactive$Source_type)), filename = file)
      
      # Write !Sample_organism
      write_append(paste0("!Sample_organism", " = ", isolate(reactive$Organism)), filename = file)

      # Write !Sample_label
      write_append(paste0("!Sample_label", " = ", isolate(reactive$Label)), filename = file)
      
      # Write !Sample_molecule
      write_append(paste0("!Sample_molecule", " = ", isolate(reactive$Molecule)), filename = file)
      
      # Write !Sample_extract_protocol
      write_append(paste0("!Sample_extract_protocol", " = ", isolate(reactive$Extraction_protocol)), filename = file)
      
      # Write !Sample_label_protocol
      write_append(paste0("!Sample_label_protocol", " = ", isolate(reactive$Labeling_protocol)), filename = file)
      
      # Write !Sample_hyb_protocol
      write_append(paste0("!Sample_hyb_protocol", " = ", isolate(reactive$Hybridization_protocol)), filename = file)
      
      # Write !Sample_scan_protocol
      write_append(paste0("!Sample_scan_protocol", " = ", isolate(reactive$Scanning_protocol)), filename = file)
      
      # Write !Sample_data_processing
      write_append(paste0("!Sample_data_processing", " = ", isolate(reactive$Data_protocol)), filename = file)
      
      # Write optionals
      if(isolate(reactive$Biomaterial_protocol) != ""){
        write_append(paste0("!Sample_biomaterial_provider_ch1", " = ", isolate(reactive$Biomaterial_protocol)), filename = file)
      }
      
      # Write optionals
      if(isolate(reactive$Treatment_protocol) != ""){
        write_append(paste0("!Sample_treatment_protocol_ch1", " = ", isolate(reactive$Treatment_protocol)), filename = file)
      }
      
      # Write optionals
      if(isolate(reactive$Growth_protocol) != ""){
        write_append(paste0("!Sample_growth_protocol_ch1", " = ", isolate(reactive$Growth_protocol)), filename = file)
      }
        
      # Subset the expression values
      normalized_sub <- subset(data()$normalized, select = i)
      
      # Iterate over each row and write the expression values
      write_append("#ID_REF =", filename = file)
      write_append(paste0("#VALUE = ", "Normalized expression counts"), filename = file)
      write_append("!Sample_table_begin", filename = file)
      write_append(paste("ID_REF", "VALUE", sep = "\t"), filename = file)
      write.table(as.matrix(normalized_sub), 
                  file = file, 
                  append = TRUE, 
                  quote = FALSE, 
                  row.names = TRUE, 
                  col.names = FALSE, 
                  sep = "\t")
      write_append("!Sample_table_end", filename = file)
    }
})
```





<style>
#sidebar.section.sidebar {
  background-color: white; 
  font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif !important;
}

.js-irs-0 .irs-bar {
border-top-color: #d01010;
border-bottom-color: #d01010;
} 

.js-irs-0 .irs-bar-edge {
border-color: #d01010;
}

.js-irs-0 .irs-single, .js-irs-0 .irs-bar-edge, .js-irs-0 .irs-bar {
background: #a00;
}

</style>



